### DISCUSSION TITLE
Ways for restricting struct to contain only value types

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#6843] [Q&A]
Is there a way in c# (hopefully in .NET)

**Comment by bernd5** [comment]
Why?

**Reply by Xyncgas** [reply]
```F#
type Purity   = PureStruct
type CERN     = Demons
type civilWar = (weapon:Purity) -> (target:CERN) -> (agent:Q) -> unit
```

**Comment by En3Tho** [comment]
Well, you can do it via generics. Basically ```struct MyType<TField1, ..TFieldN> where TField1: struct, ..```. But this has its own downsides like being unable to specify custom layout due to generics. Correct me if I'm wrong.

**Comment by spydacarnage** [comment]
You could write an analyser to do this.

**Comment by CyrusNajmabadi** [comment]
Yes. An analyzer could be trivially written to "restrict struct to contain only value types".
